File: pharos-core/src/models/tag.rs
use serde::{Deserialize, Serialize};

/// One condition of a rule. `column` is PROVENANCE — the matcher never reads
/// it. `value` is the normalized form Swift produced; `display` is the text as
/// it was captured, for the Inspector.
///
/// `kind` and `operand2` are plain optional STRINGS here, never an enum. Rust
/// does not interpret them — Swift is the only producer, exactly as it already
/// is for `tuple_key`.
///
/// An enum would reject a kind written by a newer build, and `read_tag_tuples`
/// decodes the blob with `unwrap_or_default()`, so one rejected variant loads
/// the rule with an EMPTY condition list that the next save writes back. That
/// destroys the rule silently, with no error anywhere.
///
/// `#[serde(default)]` covers an ABSENT key, which is what an old stored blob
/// has. It does nothing about an unknown VALUE; keeping the type as `String` is
/// what covers that.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TagCondition {
    pub column: String,
    /// "address" | "text" | "numeric" | "temporal" | "uuid" | "type:<name>"
    pub family: String,
    #[serde(default)]
    pub kind: Option<String>,
    pub value: String,
    #[serde(default)]
    pub operand2: Option<String>,
    pub display: String,
}

/// One tagged origin row.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TagRule {
    pub id: String,
    pub conditions: Vec<TagCondition>,
    pub tuple_key: String,
    pub origin_connection: String,
    pub origin_table: String,
    pub created_at: String,
}

/// A named indicator set.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Tag {
    pub id: String,
    pub name: String,
    /// Index into a fixed palette, not a hex string.
    pub color_index: i64,
    pub note: Option<String>,
    pub created_at: String,
    pub updated_at: String,
    pub rules: Vec<TagRule>,
}

/// A tuple as Swift sends it. The id and the timestamp are minted here.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct NewTagRule {
    pub conditions: Vec<TagCondition>,
    pub tuple_key: String,
    pub origin_connection: String,
    pub origin_table: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateTag {
    pub name: String,
    pub color_index: i64,
    pub note: Option<String>,
    pub rules: Vec<NewTagRule>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddTagRules {
    pub tag_id: String,
    pub rules: Vec<NewTagRule>,
}

/// A nil field is left as it is. A note therefore cannot be CLEARED through
/// this payload, only replaced — the Phase 5 manage sheet writes an empty
/// string for "no note".
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateTag {
    pub id: String,
    pub name: Option<String>,
    pub color_index: Option<i64>,
    pub note: Option<String>,
}

#[cfg(test)]
mod tests {
    use super::*;

    /// Rust only DEserializes the write payloads: Swift writes them. A round
    /// trip would pass with `rename_all` removed — both sides would agree with
    /// each other and disagree with Swift — so decode a literal camelCase
    /// document, exactly as Swift sends it, with EVERY optional key present.
    #[test]
    fn create_tag_decodes_swift_camel_case() {
        let json = r#"{"name":"Suspect infra","colorIndex":2,"note":"may sprint",
          "rules":[{"conditions":[{"column":"md5","family":"text","value":"d41d8c",
          "display":"D41D8C"}],"tupleKey":"K4:textV6:d41d8c",
          "originConnection":"c1","originTable":"public.certs"}]}"#;
        let create: CreateTag = serde_json::from_str(json).unwrap();
        assert_eq!(create.color_index, 2);
        assert_eq!(create.note.as_deref(), Some("may sprint"));
        assert_eq!(create.rules[0].tuple_key, "K4:textV6:d41d8c");
        assert_eq!(create.rules[0].origin_table, "public.certs");
        assert_eq!(create.rules[0].conditions[0].display, "D41D8C");
    }

    #[test]
    fn add_and_update_payloads_decode_swift_camel_case() {
        let add: AddTagRules = serde_json::from_str(
            r#"{"tagId":"t1","rules":[{"conditions":[],"tupleKey":"k","originConnection":"c",
               "originTable":"t"}]}"#,
        )
        .unwrap();
        assert_eq!(add.tag_id, "t1");

        let full: UpdateTag =
            serde_json::from_str(r#"{"id":"t1","name":"Renamed","colorIndex":4,"note":"n"}"#)
                .unwrap();
        assert_eq!(full.name.as_deref(), Some("Renamed"));
        assert_eq!(full.color_index, Some(4));

        let sparse: UpdateTag = serde_json::from_str(r#"{"id":"t1"}"#).unwrap();
        assert_eq!(sparse.name, None);
    }

    /// A SPARSE document pins the defaults and nothing more. A FULL literal
    /// document is what pins the key names — serde reports a mis-cased optional
    /// as `None`, indistinguishable from the caller omitting it.
    #[test]
    fn condition_decodes_sparse_and_full_documents() {
        let sparse: TagCondition = serde_json::from_str(
            r#"{"column":"host","family":"text","value":"abc","display":"ABC"}"#,
        )
        .unwrap();
        assert_eq!(sparse.kind, None);
        assert_eq!(sparse.operand2, None);

        let full: TagCondition = serde_json::from_str(
            r#"{"column":"port","family":"numeric","kind":"between","value":"1000",
                "operand2":"2000","display":"1000 .. 2000"}"#,
        )
        .unwrap();
        assert_eq!(full.kind.as_deref(), Some("between"));
        assert_eq!(full.operand2.as_deref(), Some("2000"));
        assert_eq!(full.family, "numeric");
        assert_eq!(full.display, "1000 .. 2000");
    }

    /// An unknown kind must round-trip verbatim. This is the test that protects
    /// a rule written by a newer build from being destroyed by an older one.
    #[test]
    fn unknown_kind_round_trips_verbatim() {
        let decoded: TagCondition = serde_json::from_str(
            r#"{"column":"h","family":"text","kind":"startsWith","value":"a","display":"a"}"#,
        )
        .unwrap();
        assert_eq!(decoded.kind.as_deref(), Some("startsWith"));
        let json = serde_json::to_string(&decoded).unwrap();
        assert!(json.contains(r#""kind":"startsWith""#), "got {}", json);
    }

    #[test]
    fn tag_serializes_camel_case() {
        let tag = Tag {
            id: "t1".into(),
            name: "Suspect infra".into(),
            color_index: 2,
            note: None,
            created_at: "2026-08-13T00:00:00Z".into(),
            updated_at: "2026-08-13T00:00:00Z".into(),
            rules: vec![TagRule {
                id: "u1".into(),
                conditions: vec![TagCondition {
                    column: "md5".into(),
                    family: "text".into(),
                    kind: None,
                    value: "d41d8c".into(),
                    operand2: None,
                    display: "D41D8C".into(),
                }],
                tuple_key: "K4:textV6:d41d8c".into(),
                origin_connection: "c1".into(),
                origin_table: "public.certs".into(),
                created_at: "2026-08-13T00:00:00Z".into(),
            }],
        };
        let json = serde_json::to_string(&tag).unwrap();
        assert!(json.contains("\"colorIndex\":2"), "got {}", json);
        assert!(json.contains("\"tupleKey\""), "got {}", json);
        assert!(json.contains("\"originConnection\""), "got {}", json);
    }
}

File: Pharos/Models/Tag.swift
import Foundation

// The Rust tag models in `pharos-core/src/models/tag.rs` all carry
// `#[serde(rename_all = "camelCase")]`, and JSONDecoder/Encoder.pharos apply NO
// key strategy, so these Swift structs use plain camelCase property names
// (matching the JSON keys exactly). Do not add snake_case CodingKeys here.
// `TagCondition` is the one type that carries CodingKeys at all, and only
// because its hand-written `init(from:)` needs them; they spell the same
// camelCase the synthesized ones would, and they list EVERY coded property.
//
// The OPPOSITE rule applies to the result models — see `ColumnDef` in
// Pharos/Models/QueryResult.swift, whose Rust structs carry no `rename_all` and
// therefore DO need CodingKeys. The two conventions sit side by side on
// purpose; a mismatch throws at run time only, so TagModelTests pins this one.

/// One condition of a rule: what to test, and what to test it against.
///
/// `value` is the NORMALIZED form and is what matching compares. `display` is
/// the text as captured or typed, byte for byte, and is what every surface
/// draws. Normalizing derives a second form beside the first; it never alters
/// what the analyst wrote.
///
/// There is deliberately NO column name here. A column never took part in
/// matching — that is what lets a hash captured under `cert_md5` resurface
/// under `certificate_hash` in another schema — and a hand-authored condition
/// has no column at all, so the FAMILY is the one description every condition
/// can share. `TagFamilyLabel` turns it into words. Row-level provenance is
/// unaffected: `originConnection` and `originTable` live on `TagRule`.
///
/// A stored blob that still carries the old `column` key decodes unchanged:
/// `CodingKeys` has no case for it, so it is ignored. No migration is needed,
/// and `TagModelTests` pins that.
///
/// The hand-written `init(from:)` is what makes `kind` and `operand2` tolerate
/// an ABSENT key, and `TagConditionKind`'s own decoder is what makes `kind`
/// tolerate an UNKNOWN VALUE. Both are needed; neither substitutes for the
/// other, and losing either destroys a rule written by a newer build.
///
/// `CodingKeys` must stay EXHAUSTIVE. Swift requires every coded property to
/// appear, and a key left out is silently dropped — the hazard
/// `memory/pharos-ffi-json-casing.md` records. There were deliberately no
/// `CodingKeys` on this type before; they exist now only because `init(from:)`
/// needs them, and they must list every property.
struct TagCondition: Codable, Equatable {
    /// "address" | "text" | "numeric" | "temporal" | "uuid" | "type:<name>"
    let family: String
    let kind: TagConditionKind
    let value: String
    /// The upper bound of a `.between`. nil for every other kind.
    ///
    /// `between` must be ONE condition rather than two comparators. Conditions
    /// in a rule are ANDed, but each is satisfied by SOME cell in the row, not
    /// the same cell — so `>= 1000` could be satisfied by one column and
    /// `<= 2000` by another, which is not a range test at all.
    let operand2: String?
    let display: String

    init(family: String, kind: TagConditionKind = .exact,
         value: String, operand2: String? = nil, display: String) {
        self.family = family
        self.kind = kind
        self.value = value
        self.operand2 = operand2
        self.display = display
    }

    private enum CodingKeys: String, CodingKey {
        case family, kind, value, operand2, display
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        family = try container.decode(String.self, forKey: .family)
        // `decodeIfPresent` covers an ABSENT key; `TagConditionKind`'s own
        // decoder covers an UNKNOWN value.
        kind = try container.decodeIfPresent(TagConditionKind.self, forKey: .kind) ?? .exact
        value = try container.decode(String.self, forKey: .value)
        operand2 = try container.decodeIfPresent(String.self, forKey: .operand2)
        display = try container.decode(String.self, forKey: .display)
    }
}

/// One tagged origin row. Its values are the tuple the matcher tests.
struct TagRule: Codable, Equatable, Identifiable {
    let id: String
    let conditions: [TagCondition]
    let tupleKey: String
    let originConnection: String
    let originTable: String
    let createdAt: String
}

/// A named indicator set: a case name, a colour, a note, and the tuples that
/// define the finding. Global — a tag carries no connection id.
struct Tag: Codable, Equatable, Identifiable {
    let id: String
    var name: String
    /// Index into `TagPalette.colors`, not a hex string.
    var colorIndex: Int
    var note: String?
    let createdAt: String
    let updatedAt: String
    var rules: [TagRule]
}

/// A tuple as Swift sends it. Rust mints the id and the timestamp.
struct NewTagRule: Codable, Equatable {
    var conditions: [TagCondition]
    /// `RuleKey.encode` is the only producer. Rust never builds one.
    var tupleKey: String
    var originConnection: String
    var originTable: String
}

struct CreateTag: Codable {
    var name: String
    var colorIndex: Int
    var note: String?
    var rules: [NewTagRule]
}

struct AddTagRules: Codable {
    var tagId: String
    var rules: [NewTagRule]
}

/// A nil field is left as it is. A note cannot be CLEARED through this payload,
/// only replaced — write an empty string for "no note".
struct UpdateTag: Codable {
    var id: String
    var name: String?
    var colorIndex: Int?
    var note: String?
}

/// One rule's conditions, replaced in place.
///
/// The point of this payload is what it does NOT carry. The rule keeps its id,
/// its `createdAt` — when the finding was first recorded — and the origin
/// columns saying where it was first observed. Editing conditions changes none
/// of those, and the delete-then-add it replaces used to reset all four.
///
/// No `tagId`: the rule id is a primary key, and the core finds the tag by
/// subquery when it bumps `updatedAt`.
struct UpdateTagRule: Codable, Equatable {
    var ruleId: String
    var conditions: [TagCondition]
    /// `RuleKey.encode` is the only producer. Rust never builds one.
    var tupleKey: String
}
